modules/tinymce/src/core/main/ts/init/Render.ts

```diff
@@ -45,9 +50,29 @@ const loadLanguage = (scriptLoader: ScriptLoader, editor: Editor) => {
   }
 };
 
-const loadTheme = (scriptLoader: ScriptLoader, editor: Editor, suffix: string): Promise<void> => {
+/**
+ * Initiate the download of both theme and model in parallel, then wait for both.
+ *
+ * Each has their own special case where nothing needs to be waited for.
+ */
+const loadThemeAndModel = (scriptLoader: ScriptLoader, editor: Editor, suffix: string): Promise<void> => {
   const theme = Options.getTheme(editor);
 
+  // Special case the 'wait for model' code if RTC is loading, as it will provide a model instead
+  let waitForModel: () => Promise<void>;
+  if (!Rtc.isRtc(editor)) {
+    const model = Options.getModel(editor);
+    const modelUrl = Options.getModelUrl(editor);
+    const url = Type.isString(modelUrl) ? editor.documentBaseURI.toAbsolute(modelUrl) : `models/${model}/model${suffix}.js`;
+    ModelManager.load(model, url).catch(() => {
+      ErrorReporter.modelLoadError(editor, url, model);
+    });
+
+    waitForModel = () => ModelManager.waitFor(model);
+  } else {
+    waitForModel = () => Promise.resolve();
+  }
```

Comment:
This really seems like the wrong approach here and in addition that, I doubt this is actually working because `Rtc.isRtc` will always return `false` at this point because `editor.plugins` is empty at this point.

Reply:
Maybe it's been moved since I last tested it. The aim is to not download the `dom` model when RTC is going to replace it anyway.

Reply:
We probably need to discuss this more as it's not clear how RTC is going to register it. It feels like the better option here would be to override or change the default model to load when RTC is enabled (and potentially the url).

The loading order also seems to be wrong here if we want plugins to control the model, as the plugins should load first and then the model in that case. We can check to see if the model is already loaded and skip loading it in that case, as it'd be loaded and registered when the RTC plugin loads.

Reply:
I don't want plugins to register a model, RTC is just a special case until the alt-model can become a core feature.

The model needs to load before plugins, so we perhaps just need a better check than `isRtc` here.

Reply:
> The model needs to load before plugins, so we perhaps just need a better check than isRtc here.

I disagree here, as the model needs to init before plugins, so the plugin can use it but it should have no need to load before them.

Reply:
You're right, I conflated them, but I don't want to restrict what plugins can do before `init` too much. I guess they're already fairly restricted so that's an ok change to make.

Reply:
oh and I forgot to clarify
> it's not clear how RTC is going to register it

That happens automatically on the TinyMCE side
https://github.com/tinymce/tinymce/blob/1f7d16e4d680742369528c9f4033fde08e7fe76f/modules/tinymce/src/core/main/ts/Rtc.ts#L316

Reply:
Okay, that's leading me even more to believe the model manager is something we shouldn't be doing then if we're going with that approach. As if we're using the ModelManager API, then RTC should be calling `ModelManager.add('rtc', () => {...})` instead of monkey patching it like that.

Reply:
Ah github didn't update with your second comment before I made mine, sorry. We probably should be discussing this in Slack.

> This seems like a bad design/idea to me as why do we even have the model manager API then if it's not going to be customisable? We might as well just chunk the files and then fork the loading based on if RTC is enabled or not.

I thought it was required if we later have two models in the core? If not then yes we can drop the manager.

Reply:
The primary goal of this code is two-fold

* Only download the DOM model if RTC isn't configured
* Download the DOM model in parallel with the theme, because both are quite large and adding an extra round-trip to the init process is highly undesirable

Reply:
We had a big discussion this morning and we settled on an approach; the `model` option will default to `plugin` when RTC is in the plugins list. This code then changes to check whether the option is set to plugin rather than a specific RTC check. I'm about to add that to the PR and then open a matching PR on the RTC side.


> But I wonder if that change needs to be it's own PR and loading the dom model could be just an empty model in the first PR then we re-open this PR with the table changes moving into that model. This PR Is massive.

yes, sorry about the size, I nearly did a model-only PR months ago when the model API seemed solid. But we didn't have anything for the model to do, and I thought the table changes might impact how the model API was structured.

For now, what you're looking for has been squashed into the first commit of the PR.

Reply:
> I wonder why the theme is loaded before the other resources. I tried tweaking it so that they load in parallel instead and from the waterfall it seems to do just that and the editor works just fine. I think it might be some legacy thing where the theme was providing apis that plugins needed on load not on init. So we might be able to change that and then also load the dom model in parallel as well.

We've logged this as TINY-8453 and will see if we can squeeze it in for 6.0

Reply:
> We had a big discussion this morning and we settled on an approach; the model option will default to plugin when RTC is in the plugins list. This code then changes to check whether the option is set to plugin rather than a specific RTC check. I'm about to add that to the PR and then open a matching PR on the RTC side.

That sounds like a viable approach until we move the model. :+1: